src/Raven.Server/Documents/Indexes/IndexStorage.cs

```diff
@@ -314,6 +329,95 @@ public static IndexType ReadIndexType(int indexId, StorageEnvironment environmen
             }
         }
 
+        public IEnumerable<string> GetReferencedCollections(RavenTransaction tx)
+        {
+            var tree = tx.InnerTransaction.ReadTree(Schema.ReferencedCollections);
+            using (var it = tree.Iterate(prefetch: true))
+            {
+                if (it.Seek(Slices.BeforeAllKeys) == false)
+                    yield break;
+
+                do
+                {
+                    yield return it.CurrentKey.ToString(); // TODO [ppekrol] ?
+                } while (it.MoveNext());
+            }
+        }
+
+        public IEnumerable<string> GetDocumentKeysThatReference(LazyStringValue key, RavenTransaction tx)
+        {
+            var reverseTree = tx.InnerTransaction.ReadTree(Schema.ReverseReferences);
+            using (var it = reverseTree.MultiRead(key))
+            {
+                if (it.Seek(Slices.BeforeAllKeys) == false)
+                    yield break;
+
+                do
+                {
+                    yield return it.CurrentKey.ToString(); // TODO [ppekrol] ?
+                } while (it.MoveNext());
+            }
+        }
+
+        public long ReadLastSeenEtagForReference(string collection, LazyStringValue key, RavenTransaction tx)
+        {
+            var tree = tx.InnerTransaction.ReadTree("$" + collection);
+
+            var value = tree?.Read(key);
+            if (value == null)
+                return 0;
+
+            return value.Reader.ReadLittleEndianInt64();
+        }
+
+        public unsafe void WriteReferences(CurrentIndexingScope indexingScope, RavenTransaction tx)
+        {
```

Comment:
Not following what is the puspose of all of those trees, seems to me they are tracking a lot of duplicated information